# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/DeviceManagerDx.cpp
# Commit: ca2e6339f35f
# Full Hash: ca2e6339f35f38cc26a203f29c8dbfeb38f96afc
# Author: Kelsey Gilbert <kelsey.gilbert@mozilla.com>
# Date: 2023-03-14 09:41:39
# Regressor Bug: 1799258
# File Overlap Count: 1
# Description:
#   Bug 1799258 - Do color-management on Windows+DComp via IDCompositionFilterEffects. r=sotaro
#   
#   + Add gfx.color_management.rec709_gamma_as_srgb:true. :'(
#   
#   In particular, rec709(16/255) -> srgb(31/255). Even though it's
# ==============================================================================

diff -r aa4e010199c6 -r ca2e6339f35f gfx/thebes/DeviceManagerDx.cpp
--- a/gfx/thebes/DeviceManagerDx.cpp	Mon Mar 13 21:04:10 2023 +0000
+++ b/gfx/thebes/DeviceManagerDx.cpp	Mon Mar 13 21:04:10 2023 +0000
@@ -30,6 +30,10 @@
 #undef NTDDI_VERSION
 #define NTDDI_VERSION NTDDI_WINBLUE
 
+// We also need this, or dcomp.h won't give us e.g. IDCompositionDevice3:
+#undef _WIN32_WINNT_WINTHRESHOLD
+#define _WIN32_WINNT_WINTHRESHOLD _WIN32_WINNT
+
 #include <d3d11.h>
 #include <dcomp.h>
 #include <ddraw.h>
@@ -50,6 +54,7 @@
 
 // It should only be used within CreateDirectCompositionDevice.
 decltype(DCompositionCreateDevice2)* sDcompCreateDevice2Fn = nullptr;
+decltype(DCompositionCreateDevice3)* sDcompCreateDevice3Fn = nullptr;
 
 // It should only be used within CreateDCompSurfaceHandle
 decltype(DCompositionCreateSurfaceHandle)* sDcompCreateSurfaceHandleFn =
@@ -117,7 +122,7 @@
   MOZ_ASSERT(gfxVars::UseWebRenderDCompWin());
 
   if (sDcompCreateDevice2Fn) {
-    return true;
+    return true;  // Already loaded.
   }
 
   nsModuleHandle module(LoadLibrarySystem32(L"dcomp.dll"));
@@ -127,6 +132,8 @@
 
   sDcompCreateDevice2Fn = (decltype(DCompositionCreateDevice2)*)GetProcAddress(
       module, "DCompositionCreateDevice2");
+  sDcompCreateDevice3Fn = (decltype(DCompositionCreateDevice3)*)GetProcAddress(
+      module, "DCompositionCreateDevice3");
   if (!sDcompCreateDevice2Fn) {
     return false;
   }
@@ -135,9 +142,6 @@
   sDcompCreateSurfaceHandleFn =
       (decltype(DCompositionCreateSurfaceHandle)*)::GetProcAddress(
           module, "DCompositionCreateSurfaceHandle");
-  if (!sDcompCreateDevice2Fn) {
-    return false;
-  }
 
   mDcompModule.steal(module);
   return true;
@@ -448,10 +452,16 @@
   HRESULT hr;
   RefPtr<IDCompositionDesktopDevice> desktopDevice;
   MOZ_SEH_TRY {
-    hr = sDcompCreateDevice2Fn(
+    hr = sDcompCreateDevice3Fn(
         dxgiDevice.get(),
         IID_PPV_ARGS(
             (IDCompositionDesktopDevice**)getter_AddRefs(desktopDevice)));
+    if (!desktopDevice) {
+      hr = sDcompCreateDevice2Fn(
+          dxgiDevice.get(),
+          IID_PPV_ARGS(
+              (IDCompositionDesktopDevice**)getter_AddRefs(desktopDevice)));
+    }
   }
   MOZ_SEH_EXCEPT(EXCEPTION_EXECUTE_HANDLER) { return; }
 